### DISCUSSION TITLE
local_offset_at on Linux multi-threaded

### DISCUSSION THREAD

**Discussion by keepsimple1** [time-rs/time#567] [Question]
Currently `local_offset_at` does not work on Linux multi-threaded programs, as explained here:

https://github.com/time-rs/time/blob/fadac54b08f99cb5628e45ec9f0d94fb9ee4932b/time/src/sys/local_offset_at/unix.rs#L31-L33

It seems that `std::env::set_var` is the problem, but which code in `timestamp_to_tm` calls `std::env::set_var`?  Or is it referring to the C function of `set_var`? 

Also: The [POSIX doc](https://pubs.opengroup.org/onlinepubs/9699919799/functions/localtime.html) seems to say `localtime_r` can provide the transformation and is thread-safe. But we still call `tzset()`. In what cases will `localtime_r` return incorrect info if we don't call `tzset()` explicitly?

**Comment by jhpratt** [comment]
The key part of the comment is this:

> This method is, on its own, safe. It is the presence of a safe, unsound way to set environment variables that makes it unsafe.

Because `std::env::set_var` is unsound (it should be unsafe but isn't), code written entirely in Rust can cause a segfault when using `time` if it weren't for the soundness check. The same is true for C's `set_var`, but that's a separate issue isn't particularly relevant here.

With regard to `localtime_r` and `tzset`, `tzset` is required because `localtime_r` is not required to call it. There's probably something more subtle that I'm forgetting, but there was definitely a reason.

**Reply by keepsimple1** [reply]
> Because `std::env::set_var` is unsound (it should be unsafe but isn't), code written entirely in Rust can cause a segfault when using `time` if it weren't for the soundness check.

Where does `time` use `std::env::set_var`? (I must have missed something, if `time` does not use `std::env::set_var`, then its problem is not relevant to `time`, no? ).

**Reply by jhpratt** [reply]
The problem is absolutely relevant to `time`, as we're the one dereferencing the pointer that is now a dangling pointer. `time` does not need to use `std::env::set_var` itself — _anyone, anywhere_ setting an environment variable _in any language_ is enough to cause the issue.
